AMDGPU: Pre-commit test to verify mode change in fp constrained operations

abhigargrepo opened the pull request:
This test will check the  mode register in case of constrained floating point operations.

abhigargrepo committed:
Pre-commit lit test

This test will check the  mode register in case of
constrained floating point operations.

arsenm reviewed (commented):
This can be an IR test. Commit message should also be more specific

abhigargrepo committed:
Merge branch 'main' into bug-siMode

abhigargrepo committed:
Pre-commit test to verify mode change in fp constrained operations

This test will check the mode register change in case of constrained floating point operations.

arsenm commented on llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:
```diff
@@ -0,0 +1,28 @@
+  tail call void @llvm.amdgcn.s.setreg(i32 2177, i32 0)
+  ret double %3
+}
+
```
Can you also add a test function using the standard llvm.set.fpenv intrinsics (and maybe set.rounding but then you depend on the PR to support that)

abhigargrepo replied:
Added a test function to check for llvm.set.fpenv intrinsics

arsenm commented on llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:
```diff
@@ -0,0 +1,28 @@
+; The si-mode-register pass is changing the default mode for FP constrained  operations.
+; It must ignore for strictfp functions.
+
+define hidden fastcc double @ignoreStrictfp(double noundef %0, double noundef %1) unnamed_addr #0 {
```
```suggestion
define double @ignoreStrictfp(double noundef %0, double noundef %1) #0 {
```

arsenm commented on llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:
```diff
@@ -0,0 +1,28 @@
+; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 2), 0
+; GCN-NEXT:    s_setpc_b64 s[30:31]
+  tail call void @llvm.amdgcn.s.setreg(i32 2177, i32 1)
+  %3 = tail call double @llvm.experimental.constrained.fadd.f64(double %0, double %1, metadata !"round.dynamic", metadata !"fpexcept.strict") #0
```
Use named values in tests

arsenm reviewed (commented):
Patch title should be made more specific for what this is a test for

abhigargrepo committed:
Merge branch 'main' into bug-siMode

abhigargrepo committed:
Pre-commit test to verify mode change in fp constrained operations

This test will check the mode register change in case of constrained floating point operations.

arsenm commented on llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:
```diff
@@ -0,0 +1,43 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -verify-machineinstrs %s -o - | FileCheck -check-prefix=GCN %s
```
```suggestion
; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 %s -o - | FileCheck -check-prefix=GCN %s
```

abhigargrepo committed:
Pre-commit test to verify mode change in fp constrained operations

This test will check the mode register change in case of constrained floating point operations.

arsenm commented on llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:
```diff
@@ -0,0 +1,43 @@
+; GCN-NEXT:    v_add_f64 v[0:1], v[0:1], v[2:3]
+; GCN-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  call void @llvm.set.fpenv.i64(i64 4)
```
Not the most realistic source value but ok

arsenm committed:
Merge branch 'main' into bug-siMode
